forgerock-openbanking-uk-aspsp-rs/forgerock-openbanking-uk-aspsp-rs-mock-store/forgerock-openbanking-uk-aspsp-rs-mock-store-server/src/main/java/com/forgerock/openbanking/aspsp/rs/store/repository/migration/v3_1_6/MongoDbPaymentsChangeLog.java

```diff
@@ -24,83 +24,215 @@
 
-        CloseableIterator<FRDomesticStandingOrderConsent3> frDomesticStandingOrderConsents = getLegacyDocuments(mongoTemplate, FRDomesticStandingOrderConsent3.class);
-        frDomesticStandingOrderConsents.forEachRemaining(d -> migrate(mongoTemplate, d, toFRDomesticStandingOrderConsent(d)));
+        int docsUpdated = getObjectsToUpdate().stream().mapToInt(objectToUpdate -> upgrade(mongoTemplate, objectToUpdate.legacyClass, objectToUpdate.writeRiskParentField)).sum();
 
-        CloseableIterator<FRDomesticStandingOrderPaymentSubmission3> frDomesticStandingOrderSubmissions = getLegacyDocuments(mongoTemplate, FRDomesticStandingOrderPaymentSubmission3.class);
-        frDomesticStandingOrderSubmissions.forEachRemaining(d -> migrate(mongoTemplate, d, toFRDomesticStandingOrderPaymentSubmission(d)));
+        elapsedTime.stop();
 
-        CloseableIterator<FRFileConsent2> frFileConsents = getLegacyDocuments(mongoTemplate, FRFileConsent2.class);
-        frFileConsents.forEachRemaining(d -> migrate(mongoTemplate, d, toFRFileConsent(d)));
+        log.info("Upgraded {} documents in {} seconds.", docsUpdated, elapsedTime.getTotalTimeSeconds());
+        log.info("-----------------------------------------------------------------------");
+        log.info("Finished migrating Payments API data from v3.1.6 to v3.1.6 patch");
+    }
 
-        CloseableIterator<FRFilePaymentSubmission2> frFilePaymentSubmissions = getLegacyDocuments(mongoTemplate, FRFilePaymentSubmission2.class);
-        frFilePaymentSubmissions.forEachRemaining(d -> migrate(mongoTemplate, d, toFRFilePaymentSubmission(d)));
+    /*
+     * Entry method to execute the upgrade bulk operation
+     * @param mongoTemplate
+     * @param legacyClazz collection class
+     * @param writeRiskParentField field name to get the write object that contains the risk object related with the collection class
+     * @param <T> generic parameter to conforms legacyClazz object
+     */
+    private <T> int upgrade(MongoTemplate mongoTemplate, Class<T> legacyClazz, String writeRiskParentField) {
+        // execution time control - start
+        StopWatch elapsedTime = start(legacyClazz);
+
+        // number of docs updated
+        int docsUpdated = 0;
+        // prepare the json path to build the criteria filter
+        String jsonPathCriteria = new StringBuffer(writeRiskParentField)
+                .append(SEPARATOR).append(RISK)
+                .append(SEPARATOR).append(DELIVERY_ADDRESS)
+                .append(SEPARATOR).append(COUNTRY_SUB_DIVISION).toString();
+
+        // Criteria to query only documents where countrySubDivision not null and obVersion is 'v3_1_6'
+        // { "${writeRiskParentField}.risk.deliveryAddress.countrySubDivision" : { "$ne" : null }, "$and" : [{ "obVersion" : "v3_1_6" }] }
+        Criteria queryCriteria = new Criteria(jsonPathCriteria)
+                .ne(null).andOperator(new Criteria(OB_VERSION_FIELD).is(OB_VERSION_VALUE));
```

Comment:
I don't think I upgraded the OB version field when I did the migration previously? I think I decided to leave it to reflect which version an object was stored against.

Therefore, you need to check if this will work for objects that were stored prior to v3.1.6.

Reply:
This works for all versions, it has been tested without the version filter but we must apply the upgrade only for objects with version 3.1.6 because the change of the datatype only have been applied since version 3.1.6.

Reply:
So what would happen if a document was stored using say v3.1.1 of the API (or whatever LBG was using prior to the latest release) and it was then migrated to v3.1.6, but this version field is still set to v3.1.1 on the document?

Also, there is only one `OBRisk1DeliveryAddress`/`FRRiskDeliveryAddress`, so if a change is made to either of them, then it affects all versions.

Reply:
Done: the filter obVersion has been deleted.

Reply:
I see you've updated your comment above, so mine no longer makes sense, but good to see you understand what I was saying now! 😆